pkg/controller/pod_iptables_eip.go

```diff
@@ -0,0 +1,360 @@
+		v = strings.Trim(v, " ")
+		if v != "" {
+			eipName = v
+		}
+	}
+	newPod := cachedPod.DeepCopy()
+	if newPod.Annotations[util.AllocatedAnnotation] != "true" {
+		err = fmt.Errorf("pod network not allocated, failed to create iptables eip %s", eipName)
+		return err
+	}
+	if _, err = c.iptablesEipsLister.Get(eipName); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return err
+		}
+		klog.V(3).Infof("handle add pod annotated iptables eip %s", eipName)
+		subnet := newPod.Annotations[util.LogicalSwitchAnnotation]
+		router := newPod.Annotations[util.LogicalRouterAnnotation]
+		natGw, err := c.getNatGw(router, subnet)
+		if err != nil {
+			klog.Errorf("failed to get vpc nat gw eip: %v", eipName, err)
+			return err
+		}
+		if err := c.createOrUpdateCrdEip(eipName, "", "", "", natGw); err != nil {
+			klog.Errorf("failed to create eip %s: %v", eipName, err)
+			return err
+		}
+	}
+	// check eip if ready and update pod eip annotation
+	var eip *kubeovnv1.IptablesEIP
+	if eip, err = c.iptablesEipsLister.Get(eipName); err != nil {
+		return err
+	}
+	// update pod eip annotation
+	if eip.Spec.V4ip != "" {
+		newPod.Annotations[util.EipAnnotation] = eip.Spec.V4ip
+		patch, err := util.GenerateStrategicMergePatchPayload(cachedPod, newPod)
+		if err != nil {
+			return err
+		}
+		if _, err := c.config.KubeClient.CoreV1().Pods(namespace).Patch(context.Background(), name,
+			types.StrategicMergePatchType, patch, metav1.PatchOptions{}, ""); err != nil {
+			if k8serrors.IsNotFound(err) {
+				return nil
+			}
+			klog.Errorf("patch pod %s/%s annotation eip failed: %v", name, namespace, err)
+			return err
+		}
+	} else {
+		return fmt.Errorf("eip %s not ready", eipName)
+	}
+	return nil
+}
+
+func (c *Controller) handleDeletePodAnnotatedIptablesEip(pod *v1.Pod) error {
+	var err error
+	var keepEipCR bool
+	klog.V(3).Infof("handle delete annotated iptables eip for pod %s/%s", pod.Namespace, pod.Name)
+
+	// keep statefulset eip cr
+	if isStateful, statefulSetName := isStatefulSetPod(pod); isStateful {
+		if !isStatefulSetDeleted(c.config.KubeClient, pod, statefulSetName) {
+			keepEipCR = true
+		}
+	}
+	eipName := PodNameToEipName(pod.Name, pod.Namespace)
+	if v, ok := pod.Annotations[util.EipNameAnnotation]; ok {
+		// keep named eip cr
+		v = strings.Trim(v, " ")
+		if v != "" {
+			keepEipCR = true
+			eipName = v
+		}
+	}
+	if keepEipCR {
+		return nil
+	}
+	klog.V(3).Infof("delete pod annotated iptables eip cr %s", eipName)
+	if err = c.config.KubeOvnClient.KubeovnV1().IptablesEIPs().Delete(context.Background(), eipName, metav1.DeleteOptions{}); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			klog.Errorf("failed to delete iptables eip %s: %v", eipName, err)
+			return err
+		}
+	}
+	return nil
+}
+
+func PodNameToEipName(name, namespace string) string {
+	return fmt.Sprintf("%s.%s", name, namespace)
+}
+
+func isStatefulSetDeleted(c kubernetes.Interface, pod *v1.Pod, statefulSetName string) bool {
```

Comment:
why not reuse isStatefulSetPodDeleted in pod.go

Reply:
this func is extract from isStatefulSetPodToDel in pod.go,
in our design, when scale down statefulset replica, we want to keep it fip, 
as isStatefulSetPodToDel  for pod it will remove pod ip when scale down.